src/GraphQL.SystemTextJson/ObjectDictionaryConverter.cs

```diff
@@ -0,0 +1,223 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace GraphQL.SystemTextJson
+{
+    /// <summary>
+    /// A custom JsonConverter for converting into a dictionary of objects of their real underlying type.
+    /// </summary>
+    /// <remarks>
+    /// With thanks to @pekkah who shared this from tanka-graphql.
+    /// </remarks>
+    public class ObjectDictionaryConverter : JsonConverter<Dictionary<string, object>>
+    {
+        private readonly JsonSerializerOptions _internalOptions = new JsonSerializerOptions()
+        {
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+        };
+
+        public override Dictionary<string, object> Read(ref Utf8JsonReader reader, Type typeToConvert,
+            JsonSerializerOptions options)
+        {
+            using var doc = JsonDocument.ParseValue(ref reader);
+            return ReadDictionary(doc.RootElement);
+        }
+
+        public override void Write(Utf8JsonWriter writer, Dictionary<string, object> value,
+            JsonSerializerOptions options)
+        {
+            JsonSerializer.Serialize(writer, value, _internalOptions);
+        }
+
+        private Dictionary<string, object> ReadDictionary(JsonElement element)
+        {
+            var result = new Dictionary<string, object>();
+            foreach (var property in element.EnumerateObject())
+            {
+                var key = property.Name;
+                var value = property.Value;
+                object resultValue = null;
+
+                switch (value.ValueKind)
+                {
+                    case JsonValueKind.Object:
+                        resultValue = ReadDictionary(value);
+                        break;
+                    case JsonValueKind.Number:
+                        resultValue = ReadNumber(value);
+                        break;
+                    case JsonValueKind.True:
+                    case JsonValueKind.False:
+                        resultValue = value.GetBoolean();
+                        break;
+                    case JsonValueKind.String:
+                        resultValue = value.GetString();
+                        break;
+                    case JsonValueKind.Null:
+                        // default value is null
+                        break;
+                    case JsonValueKind.Array:
+                        resultValue = ReadArray(value).ToList();
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Unexpected value kind: {value.ValueKind}");
+                }
+
+                result[key] = resultValue;
+            }
+
+            if (!result.Any())
```

Comment:
I propose to move `var result = new Dictionary<string, object>();` allocation from the beginning of this method:
```c#
 if (result == null)
  result = new Dictionary<string, object>();
result[key] = resultValue;
```

and remove `if (!result.Any()) return null;`

Reply:
If we can guarantee that `JsonElement` is a `ValueKind.Object`, should it always return a empty dictionary rather than `null`?

I can check upfront in the `Read` method that it is non-null and and a Object kind then remove the following:
```
            if (!result.Any())
                return null;
```

Reply:
`ReadDictionary` is called not from only `Read`.

Reply:
> should it always return a empty dictionary rather than null?

I don't know. Maybe in some cases there is a difference.

Reply:
> ReadDictionary is called not from only Read.

Yea but in the other case (when looping over internal levels) I'd imagine if it's reporting a valuekind that isn't `ValueKind.Null` and is `ValueKind.Object`, it'd have to be an object. A null check in the Read method may even be redundant as the `Utf8JsonReader` throws very quickly if it doesn't recognize anything it recognizes. 

> I don't know. Maybe in some cases there is a difference.

I think since the `.ValueKind` property is telling us it's either `Null` or an `Object` that we should match that in our internal representation.

Reply:
I think that if desired, then we can figure this issue out, but now it’s not worth it.

Reply:
I've made this change and I think it's the correct one. Going to resolve this; I'm happy with where it is and we've got bigger fish to fry.